packages/cursorless-engine/src/core/commandVersionUpgrades/generateSpokenForm.ts

```diff
@@ -0,0 +1,478 @@
+        return [modifiersExtra.last, number, pluralize(scope)];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+    }
+
+    case "range": {
+      if (
+        modifier.anchor.type === "ordinalScope" &&
+        modifier.active.type === "ordinalScope" &&
+        modifier.anchor.length === 1 &&
+        modifier.active.length === 1 &&
+        modifier.anchor.scopeType.type === modifier.active.scopeType.type
+      ) {
+        const anchor =
+          modifier.anchor.start === -1
+            ? modifiersExtra.last
+            : ordinalToSpokenForm(modifier.anchor.start + 1);
+        const active = modifierToSpokenForm(modifier.active);
+        const connective = getRangeConnective(
+          modifier.excludeAnchor,
+          modifier.excludeActive,
+        );
+        return [anchor, connective, active];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+    }
+
+    default:
+      return [modifiers[modifier.type]];
+  }
+}
+
+function relativeScopeInclusiveToSpokenForm(
+  modifier: RelativeScopeModifier,
+): RecursiveArray<string> {
+  const scope = scopeTypeToSpokenForm(modifier.scopeType);
+
+  if (modifier.length === 1) {
+    const direction =
+      modifier.direction === "forward"
+        ? connectives.forward
+        : connectives.backward;
+
+    // token forward/backward
+    return [scope, direction];
+  }
+
+  const length = numberToSpokenForm(modifier.length);
+  const scopePlural = pluralize(scope);
+
+  // two tokens
+  // This could also have been "two tokens forward"; there is no way to disambiguate.
+  if (modifier.direction === "forward") {
+    return [length, scopePlural];
+  }
+
+  // two tokens backward
+  return [length, scopePlural, connectives.backward];
+}
+
+function relativeScopeExclusiveToSpokenForm(
+  modifier: RelativeScopeModifier,
+): RecursiveArray<string> {
+  const scope = scopeTypeToSpokenForm(modifier.scopeType);
+  const direction =
+    modifier.direction === "forward" ? connectives.next : connectives.previous;
+
+  if (modifier.offset === 1) {
+    const number = numberToSpokenForm(modifier.length);
+
+    if (modifier.length === 1) {
+      // next/previous token
+      return [direction, scope];
+    }
+
+    const scopePlural = pluralize(scope);
+
+    // next/previous two tokens
+    return [direction, number, scopePlural];
+  }
+
+  if (modifier.length === 1) {
+    const ordinal = ordinalToSpokenForm(modifier.offset);
+    // second next/previous token
+    return [ordinal, direction, scope];
+  }
+
+  throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+}
+
+function scopeTypeToSpokenForm(scopeType: ScopeType): string {
+  switch (scopeType.type) {
+    case "oneOf":
+    case "customRegex":
+    case "switchStatementSubject":
+      throw new NoSpokenFormError(`Scope type '${scopeType.type}'`);
+    case "string":
+      throw new AmbiguousSpokenFormError(`Scope type '${scopeType.type}'`);
```

Comment:
Why not just throw `NoSpokenFormError` here?  I'm not sure I see the distinction.  There is no spoken form for the `string` scope, so I'd argue that `NoSpokenFormError` applies.  It used to be ambiguous, but after #1661 it is not

Reply:
Yes no it's totally fine to change that. Just wanted to be more specific in the past.